File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodemanager/NodeManagerList.java
package org.opcfoundation.webserver.addressspace.nodemanager;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.jspecify.annotations.Nullable;

import java.util.*;

public class NodeManagerList {
    private final List<NodeManagerBase> nodeManagers;
    private static int globalNsIndex = 0;

    public final static NodeManagerList nodeManagerList = new NodeManagerList();

    public static int getGlobalNsIndex()
    {
        int ret = globalNsIndex;
        globalNsIndex++;
        return ret;
    }

    public NodeManagerList()
    {
        nodeManagers = new ArrayList<>();
    }

    public List<Integer> getNsIndexes()
    {
        List<Integer> ret = new ArrayList<>();
        for (NodeManagerBase item: nodeManagers)
        {
            ret.add(item.nsIndex());
        }
        return ret;
    }

    public @Nullable NodeManagerBase getNodeManager(int nsIndex)
    {
        if (nodeManagers.size() <= nsIndex) return null;
        return nodeManagers.get(nsIndex);
    }

    public void addNodeManager(NodeManagerBase nodeManager) throws UaRuntimeException
    {
        if (nodeManager.nsIndex() != nodeManagers.size()) throw new UaRuntimeException(StatusCodes.Bad_InternalError);
        nodeManagers.add(nodeManager.nsIndex(),nodeManager);
    }
}

File: opcua-webservice-node/src/server/addressspace/nodemanager/NodeManagerList.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaError } from "opcua-webapi-ts";
import { NodeManagerBase } from "./NodeManagerBase";

export class NodeManagerList {
    private readonly _nodeManagers: Array<NodeManagerBase>;
    static readonly nodeManagerList = new NodeManagerList();

    constructor() {
        this._nodeManagers = [];
    }

    getNsIndexes(): Array<number> {
        const ret: Array<number> = [];

        for (const item of this._nodeManagers) {
            ret.push(item.nsIndex());
        }

        return ret;
    }

    getNewNsIndex(): number {
        return this._nodeManagers.length;
    }

    getNodeManager(nsIndex: number): NodeManagerBase | null {
        if (this._nodeManagers.length <= nsIndex) {
            return null;
        }

        return this._nodeManagers[nsIndex] ?? null;
    }

    addNodeManager(nodeManager: NodeManagerBase): void {
        if (nodeManager.nsIndex() !== this._nodeManagers.length) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadInternalError));
        }

        this._nodeManagers.splice(nodeManager.nsIndex(), 0, nodeManager);
    }
}